In [5]:
from scipy.cluster.hierarchy import dendrogram, linkage
from scipy.cluster.hierarchy import inconsistent
import numpy as np

In [6]:
X = [[2], [8], [0], [4], [1], [9], [9], [0]]
print('Number of samples = ', len(X))

Number of samples =  8


In [8]:
Z = linkage(X, 'ward')
Z

array([[ 5.        ,  6.        ,  0.        ,  2.        ],
       [ 2.        ,  7.        ,  0.        ,  2.        ],
       [ 0.        ,  4.        ,  1.        ,  2.        ],
       [ 1.        ,  8.        ,  1.15470054,  3.        ],
       [ 9.        , 10.        ,  2.12132034,  4.        ],
       [ 3.        , 12.        ,  4.11096096,  5.        ],
       [11.        , 13.        , 14.07183949,  8.        ]])

#### Understand Z marix

In [4]:
'''
[[  2.   7.   0.   2.] <== x[2] is linked with x[7], forming cluster x[8] = {x[2], x[7]}
 [  5.   6.   0.   2.] <== x[9]  = {x[5], x[6]}
 [  0.   4.   1.   2.] <== x[10] = {x[0], x[4]}
 [  8.  10.   1.   4.] <== x[11] = {x[8], x[10]} = {x[2], x[7], x[0], x[4]}
 [  1.   9.   1.   3.]
 [  3.  11.   2.   5.]
 [ 12.  13.   4.   8.]]
 '''

'\n[[  2.   7.   0.   2.] <== x[2] is linked with x[7], forming cluster x[8] = {x[2], x[7]}\n [  5.   6.   0.   2.] <== x[9]  = {x[5], x[6]}\n [  0.   4.   1.   2.] <== x[10] = {x[0], x[4]}\n [  8.  10.   1.   4.] <== x[11] = {x[8], x[10]} = {x[2], x[7], x[0], x[4]}\n [  1.   9.   1.   3.]\n [  3.  11.   2.   5.]\n [ 12.  13.   4.   8.]]\n '

In [9]:
# compute statistics
inconsistent(Z)

array([[0.        , 0.        , 1.        , 0.        ],
       [0.        , 0.        , 1.        , 0.        ],
       [1.        , 0.        , 1.        , 0.        ],
       [0.57735027, 0.81649658, 2.        , 0.70710678],
       [1.04044011, 1.06123822, 3.        , 1.01850858],
       [3.11614065, 1.40688837, 2.        , 0.70710678],
       [6.44583366, 6.76770586, 3.        , 1.12682288]])

A (n-1) X 4 matrix where the i’th row contains the link statistics for the non-singleton cluster i. 
- The link statistics are computed over the link heights for links  levels below the cluster i. 
- R[i,0] and R[i,1] are the mean and standard deviation of the link heights, respectively; 
- R[i,2] is the number of links included in the calculation; and 
- R[i,3] is the inconsistency coefficient,

here are 3 links included in the calculation of the fourth row:

|     Link       | Height/Distance |
| x[2] - x[7]    |   Z[0,2] = 0    |
| x[0] - x[4]    |   Z[2,2] = 1    |
| x[8] - x[10]   |   Z[3,2] = 1    |


The mean value of (1,1,0) is R[3,0] = 0.66667, and the standard deviation is R[3,1] = 0.57735 (normalized by N-1, not N). The inconsistent value is computed as: